python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll

```diff
@@ -0,0 +1,1540 @@
+/**
+ * INTERNAL: Do not use.
+ *
+ * TypeTracker based call-graph.
+ *
+ * The overall scheme for resolving calls, is to notice that Python has different kinds
+ * of callables, and resolve those with different strategies. Currently we handle these
+ * completely separately:
+ * 1. plain functions (and lambdas)
+ * 2. methods on classes
+ * 3. class instantiation
+ *
+ * So we have type-trackers for each of the 3 categories above, with some considerable
+ * effort to handle different kinds of methods on classes (staticmethod, classmethod,
+ * normal), and resolving methods correctly in regards to MRO.
+ *
+ *
+ * A goal of this library is to support modeling calls that happens by third-party
+ * libraries. For example `call_later(func, arg0, arg1, foo=val)`, and the fact that the
+ * library might inject it's own arguments, for example a context that will always be
+ * passed as the actual first argument to the function. Currently the aim is to provide
+ * enough predicates for such `call_later` function to be modeled by providing
+ * additional data-flow steps for the arguments/parameters. This means we cannot have
+ * any special logic that requires an AST call to be made before we care to figure out
+ * what callable this call might end up targeting.
+ *
+ * Specifically this means that we cannot use type-backtrackes from the function of a
+ * `CallNode`, since there is no `CallNode` to backtrack from for `func` in the example
+ * above.
+ *
+ * Note: This hasn't been 100% realized yet, so we don't currently expose a predicate to
+ * ask what targets any data-flow node has. But it's still the plan to do this!
+ */
+
+private import python
+private import DataFlowPublic
+private import DataFlowPrivate
+private import FlowSummaryImpl as FlowSummaryImpl
+private import FlowSummaryImplSpecific as FlowSummaryImplSpecific
+private import semmle.python.internal.CachedStages
+
+newtype TParameterPosition =
+  /** Used for `self` in methods, and `cls` in classmethods. */
+  TSelfParameterPosition() or
+  TPositionalParameterPosition(int pos) {
+    pos = any(Parameter p).getPosition()
```

Comment:
That's a very detailed observation, and technically correct.

In the case where the entire codebase is as follows, there will be no `ParameterNodeImpl` where `isParameterOf` holds with parameter position `0`.

```
class Foo:
    def foo(self): pass
```

If it had `n` positional arguments, we would have parameter positions `0`, `1`, ..., `n-1`. So always one too many.

Personally, I don't think it will be worth the trouble to eradicate this potential _one_ too many positional parameter position. But if you find it worthwhile I would love to learn more :)

Reply:
Nah, that's fine. Like I said, it was a strange thought. Thanks for the thorough explanation! 🙂